analysis/simulation.py

```diff
@@ -0,0 +1,91 @@
+"""Functions for simulating various PFE methods across a data set."""
+
+import collections
+import itertools
+
+GraphTotals = collections.namedtuple(
+    "GraphTotals", ["time_ms", "request_bytes", "response_bytes"])
+
+# Bandwidth is in bytes per ms, RTT is ms
+NetworkModel = collections.namedtuple(
+    "NetworkModel", ["name", "rtt", "bandwidth_up", "bandwidth_down"])
+
+
+class GraphHasCyclesError(Exception):
+  """Encountered a graph that can't be completed because
+  it contains cycles."""
+
+
+def network_time_for(requests, network_model):
+  """Returns the time needed to execute a graph of requests under a given network model."""
+  request_bytes = sum(request.request_size for request in requests)
+  response_bytes = sum(response.response_size for response in requests)
+  time = (network_model.rtt + request_bytes / network_model.bandwidth_up +
+          response_bytes / network_model.bandwidth_down)
+  return (time, request_bytes, response_bytes)
+
+
+def simulate_all(sequences, pfe_methods, network_models, font_directory):
+  """Simulate the matrix of {sequences} x {pfe_methods} x {network_models}.
+
+  For each element compute a set of summary metrics, total time, total
+  request bytes sent, and total response bytes sent.
+  """
+  result = dict()
+  for args in itertools.product(sequences, pfe_methods, network_models):
+    key = "%s (%s)" % (args[1].name(), args[2].name)
```

Comment:
no real reason to stringify, tuples are hashable so just use (name1, name2) as the key directly

Reply:
This is fixed in a future PR.